url7b90c1010b_DastanIqbal_HelloZMQwithJava_tgz-pJ8-ClientJ8/src/zmq/Poller.java

```diff
@@ -0,0 +1,277 @@
+/*
+    Copyright (c) 2009-2011 250bpm s.r.o.
+    Copyright (c) 2007-2009 iMatix Corporation
+    Copyright (c) 2007-2011 Other contributors as noted in the AUTHORS file
+
+    This file is part of 0MQ.
+
+    0MQ is free software; you can redistribute it and/or modify it under
+    the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    0MQ is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+package zmq;
+
+import java.io.IOException;
+import java.nio.channels.CancelledKeyException;
+import java.nio.channels.ClosedChannelException;
+import java.nio.channels.SelectableChannel;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.Selector;
+import java.util.HashMap;
+import java.util.Iterator;
+import java.util.Map;
+
+public class Poller extends PollerBase implements Runnable {
+
+    private static class PollSet {
+        protected IPollEvents handler;
+        protected SelectionKey key;
+        protected int ops;
+        protected boolean cancelled;
+        
+        protected PollSet(IPollEvents handler) {
+            this.handler = handler;
+            key = null;
+            cancelled = false;
+            ops = 0;
+        }
+    }
+    //  This table stores data for registered descriptors.
+    final private Map<SelectableChannel, PollSet> fd_table;
+
+    //  If true, there's at least one retired event source.
+    private boolean retired;
+
+    //  If true, thread is in the process of shutting down.
+    volatile private boolean stopping;
+    volatile private boolean stopped;
+    
+    private Thread worker;
+    private Selector selector;
+    final private String name;
+    
+    public Poller() {
+        this("poller");
+    }
+    
+    public Poller(String name_) {
+        
+        name = name_;
+        retired = false;
+        stopping = false;
+        stopped = false;
+        
+        fd_table = new HashMap<SelectableChannel, PollSet>();
+        try {
+            selector = Selector.open();
+        } catch (IOException e) {
+            throw new ZError.IOException(e);
+        }
+    }
+
+    public void destroy() {
+        
+        if (!stopped) {
+            try {
+                worker.join();
+            } catch (InterruptedException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)